src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,385 @@
+ */
+
+#include "cluster/partition_balancer_planner.h"
+
+#include "cluster/health_monitor_types.h"
+#include "cluster/scheduling/constraints.h"
+#include "cluster/scheduling/types.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+
+#include <utility>
+
+namespace cluster {
+
+partition_balancer_planner::partition_balancer_planner(
+  ss::sharded<topic_table>& topic_table,
+  ss::sharded<partition_allocator>& partition_allocator,
+  planner_config&& config)
+  : _config(std::move(config))
+  , _topic_table(topic_table)
+  , _partition_allocator(partition_allocator) {}
+
+void partition_balancer_planner::clear_status() {
+    _ntp_sizes.clear();
+    _node_reports.clear();
+    _assigned_reallocation_sizes.clear();
+    _unavailable_nodes.clear();
+    _nodes_with_disk_constraints_violation.clear();
+    _cur_batch_size = 0;
+}
+
+void partition_balancer_planner::
+  init_ntp_sizes_and_node_reports_from_health_report(
+    const cluster_health_report& health_report) {
+    for (const auto& node_report : health_report.node_reports) {
+        _node_reports[node_report.id] = node_report;
+        for (const auto& tp_ns : node_report.topics) {
+            for (const auto& partition : tp_ns.partitions) {
+                _ntp_sizes[model::ntp(
+                  tp_ns.tp_ns.ns, tp_ns.tp_ns.tp, partition.id)]
+                  = partition.size_bytes;
+            }
+        }
+    }
+    return;
+}
+
+static bool
+node_disk_space_sorter(node_disk_space const& lhs, node_disk_space const& rhs) {
+    return lhs.free_space_percent < rhs.free_space_percent;
+}
+
+void partition_balancer_planner::
+  calculate_nodes_with_disk_constraints_violation() {
+    for (const auto& n : _node_reports) {
+        for (const auto& disk : n.second.local_state.disks) {
+            double available_space = double(disk.free) / double(disk.total);
+            if (available_space < double(1) - _config.max_disk_usage()) {
+                _nodes_with_disk_constraints_violation.emplace_back(
+                  node_disk_space{
+                    .node_id = n.second.id,
+                    .free_space_percent = available_space,
+                    .free_space = disk.free,
+                    .total_space = disk.total});
+            }
+        }
+    }
+    std::sort(
+      _nodes_with_disk_constraints_violation.begin(),
+      _nodes_with_disk_constraints_violation.end(),
+      &node_disk_space_sorter);
+}
+
+void partition_balancer_planner::calculate_unavailable_nodes(
+  const std::vector<raft::follower_metrics>& follower_metrics) {
+    const auto now = raft::clock_type::now();
+    for (const auto& follower : follower_metrics) {
+        if (
+          now - follower.last_heartbeat
+          > _config.node_availability_timeout_sec()) {
+            _unavailable_nodes.insert(follower.id);
+        }
+    }
+}
+
+std::vector<model::broker_shard>
+partition_balancer_planner::get_stable_replicas(
```

Comment:
As discussed offline, we need to allow moves that leave some of the replicas that are on nodes violating constraints intact, but improve situation somewhat. Examples:
* Suppose in a 5-node cluster nodes 1 and 5 are empty, nodes 2,3,4 are full. Currently we can't schedule any movements in this cluster because one of the replicas will always be on a full node. But moving some replicas away from the full nodes (even if some of the replicas in the replica group remain on the full nodes) can unblock further movements.
* Suppose in the same cluster additionally node 5 is unavailable. We need to be able to move replicas from node 5 to node 1.